### PR TITLE
Fix + spec for #345.

### PR DESCRIPTION
**Pull request by dblock**
Fix from @fedenusy for #345 + a spec.

### CODE REVIEW

**Comment by fedenusy** [comment]
Thanks man, my test writing's a little rusty.

I actually saw a couple `ConnectionPool::PoolShuttingDownError`s last night, so this isn't a full-proof fix. But it definitely helps. I'll think about it some more later today.

**Comment by fedenusy** [comment]
I think the full-proof solution involves creating `@down_mutex` when the node gets initialized. Then:

```
def down?
  @down_mutex.synchronize { @down_at }
end

def down!
  @down_mutex.synchronize { @down_at = Time.new }
  @pool = nil
  @latency = nil
  Connection::Manager.shutdown(self)
end
```

What do you think @dblock?

File: lib/moped/node.rb [hunk 1]
```diff
@@ -150,10 +150,10 @@ def disconnect
     #
     # @since 2.0.0
     def down!
-      @pool = Connection::Manager.shutdown(self)
       @down_at = Time.new
+      @pool = nil
       @latency = nil
-      true
+      Connection::Manager.shutdown(self)
```
Comment (arthurnn) [inline lib/moped/node.rb]: we should still return `true` here , and not nil.

**Comment by arthurnn** [comment]
One small comment, otherwise I will merge after

File: lib/moped/node.rb (hunk 1 above)
Comment (fedenusy) [inline lib/moped/node.rb]: Docstring says `# @return [ nil ] Nothing.`, so we should update that as well.

Comment (arthurnn) [inline lib/moped/node.rb]: good point... if doc says that, I think it is fine to return `nil`

**Comment by arthurnn** [comment]
thanks

**Comment by dblock** [comment]
@fedenusy Clearly you need a lock to decide whether to call `shutdown` again, so it's a bit more than about setting `down_at`. I think you should try to PR that.

**Comment by fedenusy** [comment]
Also anything that calls `down?` needs to to perform subsequent logic with a mutex that prevents `shutdown`. Might get pretty hairy.

Hopefully this fix is good enough.

**Comment by sahin** [comment]
@arthurnn is this released and also added to mongoid ?

**Comment by arthurnn** [comment]
@sahin not yet released.

**Comment by sahin** [comment]
@arthurnn, it can be so good to have mongoid 4.0.3 with just a bump to the new moped

**Comment by steve-rodriguez** [comment]
@durran @arthurnn @fedenusy, when will a 2.0.4 or 2.1 version of Moped be released that contains the latest PRs to address all the primary/slave and connection pool issues?  Thx!

**Comment by dblock** [comment]
Fyi, been running this well in production at [artsy.net](https://www.artsy.net) for the last 72 hours.

**Comment by sahin** [comment]
+1

**Comment by arthurnn** [comment]
https://rubygems.org/gems/moped/versions/2.0.4 released

**Comment by sahin** [comment]
@arthurnn , thx a lot.  

can u also update mongoid ?

**Comment by arthurnn** [comment]
We should not need to update mongoid, as it uses Moped 2.0.x : https://github.com/mongoid/mongoid/blob/master/mongoid.gemspec#L24

a `bundle update moped` should be enough in your rails app.

**Comment by sahin** [comment]
@arthurnn of course, we will do "bundle update moped" 

but then only the people who is aware of the issue will have fix.

**Comment by InvisibleMan** [comment]
It isn't fixed.

It happens in multiple threads.
MOPED use wrong thread-safe code.
Simple test:

mongoid.yml:
```yml
development:
  sessions:
    default:
      database: test_development
      hosts:
        - localhost:27017
      options:
  options:
    preload_models: true
```
moped_multi_thread_fail_test.rb:
```ruby
require 'mongoid'

NUMBER_THREADS = 500
ENV["MONGOID_ENV"] = 'development'

Mongoid.load!('mongoid.yml')

class Test
  include Mongoid::Document

  field :cnt, type: Integer, default: 0
end

def multi_thread_update_counters(id, thread_count)
  threads = []
 
  thread_count.times do |i|
    threads[i] = Thread.new do
      Test.collection.find({_id: id}).update({'$inc' => {'cnt' => 1}}, {safe: true, upsert: true})
    end
  end
  threads.each {|t| t.join }
end

multi_thread_update_counters(Test.create.id, NUMBER_THREADS)
```

use:

```$: ruby moped_multi_thread_fail_test.rb```

And 'ConnectionPool::PoolShuttingDownError':
```
/home/user/.rvm/gems/ruby-2.2.1/gems/connection_pool-2.1.1/lib/connection_pool/timed_stack.rb:78:in `block (2 levels) in pop': ConnectionPool::PoolShuttingDownError (ConnectionPool::PoolShuttingDownError)
	from /home/user/.rvm/gems/ruby-2.2.1/gems/connection_pool-2.1.1/lib/connection_pool/timed_stack.rb:77:in `loop'
	from /home/user/.rvm/gems/ruby-2.2.1/gems/connection_pool-2.1.1/lib/connection_pool/timed_stack.rb:77:in `block in pop'
	from /home/user/.rvm/gems/ruby-2.2.1/gems/connection_pool-2.1.1/lib/connection_pool/timed_stack.rb:76:in `synchronize'
	from /home/user/.rvm/gems/ruby-2.2.1/gems/connection_pool-2.1.1/lib/connection_pool/timed_stack.rb:76:in `pop'
	from /home/user/.rvm/gems/ruby-2.2.1/gems/connection_pool-2.1.1/lib/connection_pool.rb:78:in `checkout'
	from /home/user/.rvm/gems/ruby-2.2.1/gems/connection_pool-2.1.1/lib/connection_pool.rb:60:in `with'
	from /home/user/.rvm/gems/ruby-2.2.1/gems/moped-2.0.4/lib/moped/node.rb:114:in `connection'
	from /home/user/.rvm/gems/ruby-2.2.1/gems/moped-2.0.4/lib/moped/node.rb:140:in `disconnect'
	from /home/user/.rvm/gems/ruby-2.2.1/gems/moped-2.0.4/lib/moped/failover/disconnect.rb:26:in `execute'
	from /home/user/.rvm/gems/ruby-2.2.1/gems/moped-2.0.4/lib/moped/node.rb:185:in `rescue in ensure_connected'
	from /home/user/.rvm/gems/ruby-2.2.1/gems/moped-2.0.4/lib/moped/node.rb:188:in `ensure_connected'
	from /home/user/.rvm/gems/ruby-2.2.1/gems/moped-2.0.4/lib/moped/node.rb:589:in `block in flush'
	from /home/user/.rvm/gems/ruby-2.2.1/gems/moped-2.0.4/lib/moped/node.rb:617:in `block in logging'
	from /home/user/.rvm/gems/ruby-2.2.1/gems/activesupport-4.2.0/lib/active_support/notifications.rb:164:in `block in instrument'
	from /home/user/.rvm/gems/ruby-2.2.1/gems/activesupport-4.2.0/lib/active_support/notifications/instrumenter.rb:20:in `instrument'
	from /home/user/.rvm/gems/ruby-2.2.1/gems/activesupport-4.2.0/lib/active_support/notifications.rb:164:in `instrument'
	from /home/user/.rvm/gems/ruby-2.2.1/gems/moped-2.0.4/lib/moped/instrumentable.rb:31:in `instrument'
	from /home/user/.rvm/gems/ruby-2.2.1/gems/moped-2.0.4/lib/moped/node.rb:616:in `logging'
	from /home/user/.rvm/gems/ruby-2.2.1/gems/moped-2.0.4/lib/moped/node.rb:587:in `flush'
	from /home/user/.rvm/gems/ruby-2.2.1/gems/moped-2.0.4/lib/moped/node.rb:391:in `process'
	from /home/user/.rvm/gems/ruby-2.2.1/gems/moped-2.0.4/lib/moped/operation/read.rb:48:in `execute'
	from /home/user/.rvm/gems/ruby-2.2.1/gems/moped-2.0.4/lib/moped/node.rb:648:in `read'
	from /home/user/.rvm/gems/ruby-2.2.1/gems/moped-2.0.4/lib/moped/node.rb:90:in `command'
	from /home/user/.rvm/gems/ruby-2.2.1/gems/moped-2.0.4/lib/moped/node.rb:432:in `refresh'
	from /home/user/.rvm/gems/ruby-2.2.1/gems/moped-2.0.4/lib/moped/cluster.rb:188:in `block in refresh'
	from /home/user/.rvm/gems/ruby-2.2.1/gems/moped-2.0.4/lib/moped/cluster.rb:200:in `each'
	from /home/user/.rvm/gems/ruby-2.2.1/gems/moped-2.0.4/lib/moped/cluster.rb:200:in `refresh'
	from /home/user/.rvm/gems/ruby-2.2.1/gems/moped-2.0.4/lib/moped/cluster.rb:151:in `nodes'
	from /home/user/.rvm/gems/ruby-2.2.1/gems/moped-2.0.4/lib/moped/cluster.rb:246:in `with_primary'
	from /home/user/.rvm/gems/ruby-2.2.1/gems/moped-2.0.4/lib/moped/query.rb:426:in `update'
	from /home/user/.rvm/gems/ruby-2.2.1/gems/mongoid-4.0.2/lib/mongoid/query_cache.rb:117:in `update_with_clear_cache'
```

**Pull request merged**